presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java

```diff
@@ -107,6 +107,14 @@ public void testProject()
         assertQuery("SELECT column_name, data_type FROM tpch.information_schema.columns ORDER BY 1 DESC, 2 DESC LIMIT 1", "VALUES ('with_hierarchy', 'varchar')");
     }
 
+    @Test
+    public void testLimit()
+    {
+        assertQuery("SELECT count(*) FROM (SELECT * from tpch.information_schema.columns LIMIT 1)", "VALUES 1");
```

Comment:
can we count metadata calls to prove that metadata is called less times with lower limit?

Reply:
Yes. I added two tests to `TestInformationSchemaConnector::testMetadataCalls`.

Also, through doing this, I found that we can make `InformationSchemaPageSource::addColumnsRecordsFor` finer-grained to reduce the number of metadata calls for queries with a limit clause. Instead of calling `listTableColumns` directly with a schema prefix, we can call `listTables` firstly and then call `listTableColumns` for each table prefix. In this way, after a limit is hit, a schema's remaining tables won't be enumerated.  But this approach makes code messier and won't help a lot if the table distribution among schemas is not skewed. So I'd rather not do this. Please let me know if you feel like it's better to do so.

Reply:
> I don't see them.

Added now.

> Are there any other disadvantages other than code more complicated? I think if we can optimize some real life use case then I think it is worth doing so. Maybe it could be done after this PR with some additional abstraction or refactoring to handle this code complication.

I tried to implement this finer-grained approach and found that it does incur inefficiency in terms of the number of metastore calls. For Hive Connector, `listTableColumns` performs better on schema prefixes than table prefixes. In other words, it increases the number of metastore calls to unfold a schema prefix and then perform `listTableColumns` on each table prefix. So I didn't use this approach in the updated PR. But as a basis for discussion, the finer-grained approach I'm talking about is like:
```Java
private void addColumnsRecords(QualifiedTablePrefix prefix)
{
    List<QualifiedTablePrefix> tablePrefixes;
    if (prefix.getTableName().isPresent()) {
        tablePrefixes = ImmutableList.of(prefix);
    }
    else {
        tablePrefixes = Stream.concat(
                metadata.listTables(session, prefix).stream(),
                metadata.listViews(session, prefix).stream())
                .map(QualifiedObjectName::asQualifiedTablePrefix)
                .collect(toImmutableList());
    }

    for (QualifiedTablePrefix tablePrefix : tablePrefixes) {
        for (Map.Entry<SchemaTableName, List<ColumnMetadata>> entry : listTableColumns(session, metadata, accessControl, tablePrefix).entrySet()) {
            SchemaTableName tableName = entry.getKey();
            int ordinalPosition = 1;
            for (ColumnMetadata column : entry.getValue()) {
                if (column.isHidden()) {
                    continue;
                }
                addRecord(
                        prefix.getCatalogName(),
                        tableName.getSchemaName(),
                        tableName.getTableName(),
                        column.getName(),
                        ordinalPosition,
                        null,
                        "YES",
                        column.getType().getDisplayName(),
                        column.getComment(),
                        column.getExtraInfo(),
                        column.getComment());
                ordinalPosition++;
                if (isLimitExhausted()) {
                    return;
                }
            }
        }
    }
}
```